packages/adapters-library/src/adapters/aave-v3/products/rewards/aaveV3RewardsAdapter.ts

```diff
@@ -0,0 +1,391 @@
+    [Chain.Base]: getAddress('0xf9cc4F0D883F1a1eb2c253bdb46c254Ca51E1F44'),
+    [Chain.Polygon]: getAddress('0x929EC64c34a17401F460460D4B9390518E5B473e'),
+    [Chain.Optimism]: getAddress('0x929EC64c34a17401F460460D4B9390518E5B473e'),
+    [Chain.Linea]: '', // not deployed on linea
+  }
+
+  private INCENTIVES_CONTRACT_DETAILS: Erc20Metadata
+
+  private provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+    helpers,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+
+    this.incentivesContract = IncentivesContract__factory.connect(
+      this.INCENTIVES_CONTRACT_ADDRESSES[this.chainId],
+      this.provider,
+    )
+
+    /**
+     * Fake protocol token created to satisfy return type
+     */
+    this.INCENTIVES_CONTRACT_DETAILS = {
+      address: this.INCENTIVES_CONTRACT_ADDRESSES[this.chainId],
+      name: 'Aave V3 Rewards',
+      symbol: 'Rewards',
+      decimals: 18,
+    }
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'AaveV3',
+      description: 'AaveV3 defi adapter',
+      siteUrl: 'https:',
+      iconUrl: AAVE_ICON_URL,
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+      productId: this.productId,
+    }
+  }
+
+  /**
+   *
+   * AToken addresses required to query reward balances
+   */
+  async getAaveTokenAddresses(): Promise<string[]> {
+    const [aTokens, stableDebtTokens, variableDebtTokens] = await Promise.all([
+      this.helpers.metadataProvider.getMetadata({
+        protocolId: Protocol.AaveV3,
+        productId: 'a-token',
+      }),
+      this.helpers.metadataProvider.getMetadata({
+        protocolId: Protocol.AaveV3,
+        productId: 'stable-debt-token',
+      }),
+      this.helpers.metadataProvider.getMetadata({
+        protocolId: Protocol.AaveV3,
+        productId: 'variable-debt-token',
+      }),
+    ])
+    return [
+      ...aTokens.map((token) => token.address),
+      ...stableDebtTokens.map((token) => token.address),
+      ...variableDebtTokens.map((token) => token.address),
+    ]
+  }
+
+  @CacheToDb
+  async getProtocolTokens(): Promise<ProtocolToken[]> {
+    const rewardTokenAddresses = await this.incentivesContract.getRewardsList()
+
+    const rewardTokens = await Promise.all(
+      rewardTokenAddresses.map((rewardToken) => {
+        return this.helpers.getTokenMetadata(rewardToken)
+      }),
+    )
+
+    return [
+      { ...this.INCENTIVES_CONTRACT_DETAILS, underlyingTokens: rewardTokens },
+    ]
+  }
+
+  private detectPositionEventSignature(
+    eventSignature = 'Transfer(address,address,uint256)',
+  ): string {
+    return ethers.id(eventSignature)
+  }
```

Comment:
@bergarces please internalise this - I think we need to make it part of the IProtocolAdapter so that we can make all adapters eth-event driven

Reply:
Do you mean to use it when we cannot take advantage of the one we run automatically outside the adapter?

An idea here would be to create a new public method that returns open positions based on events and use the questionare answers to see if we can populate it with a helper (which is ok for most tokens). Then we can always check if it's implemented and use it outside the adapter, instead of relying on the developer to do the filtering.

Reply:
I mean instead of running it outside the adapter we move it to inside the adapters. 

Many will be Transfer(address, address, amount)